### ISSUE TITLE
[QUESTION] Need suggestion regarding Connection reset issue while connecting to Always On from Java docker container.

### ISSUE THREAD

**Issue by udaykiranraju1983** [microsoft/mssql-jdbc#1384]
## Question
Need help resolving the below issue.

**Issue :** 
Getting connection reset issue while connecting to the always On MS SQL server from stand alone java application deployed in java docker container.

Note : Stand alone application is working fine from windows but not working from docker image deployed on linux.

**Java code in stand alone App :**

public class SQLDatabaseConnection {
   
    public static void main(String[] args) {
        String connectionUrl =

                "jdbc:sqlserver://SXXXXXXX05E:49001;"
                        + "database=AdventureWorks;"
                        + "user=yourusername@yourserver;"
                        + "password=yourpassword;"
                        + "loginTimeout=30;"
                        + "multisubnetfailover=true";
        try (Connection connection = DriverManager.getConnection(connectionUrl);) {
             // Code here.
        }
        catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

**Driver version:** : mssql-jdbc-8.2.2.jre8.jar
  **Docker Java base image**  : openjdk 8 
Java docker container is deployed on linux node using kubernetes.

SQL server listener(SXXXXXXX05E) (always on) has two IP addreses-> 10.100.200.111 – offline and 10.100.200.123 is online 

**Observations :**  with MultisubnetfailOver=true and timeout as 30 secs , it is trying to connect 3 times. 

The main issue I am seeing is 
 When two threads are launched to get connection, socket connector for **offline IP address** results getting as SUCCESS(see ->The socket of the following thread has been chosen:SocketConnector:9) then immediately getting connection reset. I
 I thought offline IP address should throw timeout and only socket connector with Active IP should be choosen.

Please correct me if I am wrong in my assumptions and help me in fixing the issue.

**Log**  

Failed Scenario :
[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 created by (SQLServerDriver:1)

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 Start time: 1590790973388 Time out time: 1590791003388 Timeout Unit Interval: 2400

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt server name: SXXXXXXX05E port: 49001 InstanceName: null useParallel: true

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt endtime: 1590790975788

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt No: 0

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 Connecting with server: SXXXXXXX05E port: 49001 Timeout slice: 2400 Timeout Full: 30

[           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Opening TCP socket...

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) Total no of InetAddresses: 2. They are: SXXXXXXX05E /10.100.200.111;SXXXXXXX05E /10.100.200.123;

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3))Using Threading with timeout:2400

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) TimeRemaining:2400; Result:UNKNOWN; Max. open thread count: 2; Current open thread count:0

[pool-6-thread-1] c.m.s.j.i.SocketConnector                : **SocketConnector:9**(SocketFinder(TDSChannel (ConnectionID:3))) connecting to InetSocketAddress:SXXXXXXX05E /10.100.200.111:49001 with timeout:2400

[pool-6-thread-2] c.m.s.j.i.SocketConnector                : SocketConnector:10(SocketFinder(TDSChannel (ConnectionID:3))) connecting to InetSocketAddress:SXXXXXXX05E /10.100.200.123:49001 with timeout:2400

[pool-6-thread-1] c.m.s.j.i.SocketFinder                   : The following child thread is waiting for socketFinderLock:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-1] c.m.s.j.i.SocketFinder                   : The following child thread acquired socketFinderLock:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-1] c.m.s.j.i.SocketFinder                   : **The socket of the following thread has been chosen:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))**

[pool-6-thread-1] c.m.s.j.i.SocketFinder                   : The following child thread is waiting for parentThreadLock:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread is waiting for socketFinderLock:SocketConnector:10(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-1] c.m.s.j.i.SocketFinder                   : The following child thread acquired parentThreadLock:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-1] c.m.s.j.i.SocketFinder                   : The following child thread released parentThreadLock and notified the parent thread:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-1] c.m.s.j.i.SocketFinder                   : The following child thread released socketFinderLock:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread acquired socketFinderLock:SocketConnector:10(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread is waiting for parentThreadLock:SocketConnector:10(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread acquired parentThreadLock:SocketConnector:10(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread released parentThreadLock and notified the parent thread:SocketConnector:10(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread released socketFinderLock:SocketConnector:10(SocketFinder(TDSChannel (ConnectionID:3)))

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) The parent thread wokeup.

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) TimeRemaining:2398; Result:SUCCESS; Max. open thread count: 2; Current open thread count:0

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)): Closing TCP socket:Socket[addr=SXXXXXXX05E /10.100.200.123,port=49001,localport=36682]

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 ClientConnectionId: d9091629-33ce-4034-8dcc-bc86a2da0ee6 Requesting encryption level:OFF

[           main] c.m.s.j.i.T.Channel                      : **TDSChannel (ConnectionID:3) read failed:Connection reset**

[           main] c.m.s.j.i.SQLServerException             : *** SQLException:ConnectionID:3 ClientConnectionId: d9091629-33ce-4034-8dcc-bc86a2da0ee6 com.microsoft.sqlserver.jdbc.SQLServerException: Connection reset ClientCon

[           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Closing TCP socket...

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 ClientConnectionId: d9091629-33ce-4034-8dcc-bc86a2da0ee6 Prelogin error: host SXXXXXXX05E  port 49001 Error reading prelogin response: Connection

[           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Closing TCP socket...

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 ClientConnectionId: d9091629-33ce-4034-8dcc-bc86a2da0ee6 sleeping milisec: 100

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt server name: SXXXXXXX05E  port: 49001 InstanceName: null useParallel: true

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt endtime: 1590790988398

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt No: 1

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 Connecting with server: SXXXXXXX05E port: 49001 Timeout slice: 4800 Timeout Full: 30

[           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Opening TCP socket...

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) Total no of InetAddresses: 2. They are: SXXXXXXX05E /10.100.200.111;SXXXXXXX05E /10.100.200.123;

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3))Using Threading with timeout:4800

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) TimeRemaining:4800; Result:UNKNOWN; Max. open thread count: 2; Current open thread count:0

[pool-6-thread-2] c.m.s.j.i.SocketConnector                : SocketConnector:25(SocketFinder(TDSChannel (ConnectionID:3))) connecting to InetSocketAddress:SXXXXXXX05E /10.100.200.111:49001 with timeout:4800

[pool-6-thread-1] c.m.s.j.i.SocketConnector                : SocketConnector:26(SocketFinder(TDSChannel (ConnectionID:3))) connecting to InetSocketAddress:SXXXXXXX05E /10.100.200.123:49001 with timeout:4800

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread is waiting for socketFinderLock:SocketConnector:25(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread acquired socketFinderLock:SocketConnector:25(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The socket of the following thread has been chosen:SocketConnector:25(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread is waiting for parentThreadLock:SocketConnector:25(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread acquired parentThreadLock:SocketConnector:25(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread released parentThreadLock and notified the parent thread:SocketConnector:25(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread released socketFinderLock:SocketConnector:25(SocketFinder(TDSChannel (ConnectionID:3)))

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) The parent thread wokeup.

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) TimeRemaining:4800; Result:SUCCESS; Max. open thread count: 2; Current open thread count:1

[           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)): Closing TCP socket:Socket[addr=SXXXXXXX05E /10.100.200.123,port=49001,localport=37168]

[           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 ClientConnectionId: 545d8dc6-dcd8-41f5-97b2-f85523cd92c1 Requesting encryption level:OFF

 [           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3) read failed:Connection reset

 [           main] c.m.s.j.i.SQLServerException             : *** SQLException:ConnectionID:3 ClientConnectionId: 545d8dc6-dcd8-41f5-97b2-f85523cd92c1 com.microsoft.sqlserver.jdbc.SQLServerException: Connection reset ClientCo

[           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Closing TCP socket...

 [           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 ClientConnectionId: 545d8dc6-dcd8-41f5-97b2-f85523cd92c1 Prelogin error: host SXXXXXXX05E  port 49001 Error reading prelogin response: Connection

[           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Closing TCP socket...

 [           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 ClientConnectionId: 545d8dc6-dcd8-41f5-97b2-f85523cd92c1 sleeping milisec: 200

 [           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt server name: SXXXXXXX05E port: 49001 InstanceName: null useParallel: true

 [           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt endtime: 1590791001001

 [           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 This attempt No: 2

 [           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 Connecting with server: SXXXXXXX05E port: 49001 Timeout slice: 7199 Timeout Full: 30

 [           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Opening TCP socket...

 [           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) Total no of InetAddresses: 2. They are: SXXXXXXX05E/10.100.200.111;SXXXXXXX05E/10.100.200.123;

 [           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3))Using Threading with timeout:7199

 [           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) TimeRemaining:7199; Result:UNKNOWN; Max. open thread count: 2; Current open thread count:0

 [pool-6-thread-2] c.m.s.j.i.SocketConnector                : SocketConnector:43(SocketFinder(TDSChannel (ConnectionID:3))) connecting to InetSocketAddress:SXXXXXXX05E/10.100.200.111:49001 with timeout:7199

 [pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread is waiting for socketFinderLock:SocketConnector:43(SocketFinder(TDSChannel (ConnectionID:3)))

 [pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread acquired socketFinderLock:SocketConnector:43(SocketFinder(TDSChannel (ConnectionID:3)))

 [pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The socket of the following thread has been chosen:SocketConnector:43(SocketFinder(TDSChannel (ConnectionID:3)))

 [pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread is waiting for parentThreadLock:SocketConnector:43(SocketFinder(TDSChannel (ConnectionID:3)))

 [pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread acquired parentThreadLock:SocketConnector:43(SocketFinder(TDSChannel (ConnectionID:3)))

 [pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread released parentThreadLock and notified the parent thread:SocketConnector:43(SocketFinder(TDSChannel (ConnectionID:3)))

 [pool-6-thread-2] c.m.s.j.i.SocketFinder                   : The following child thread released socketFinderLock:SocketConnector:43(SocketFinder(TDSChannel (ConnectionID:3)))

 [pool-6-thread-1] c.m.s.j.i.SocketConnector                : SocketConnector:44(SocketFinder(TDSChannel (ConnectionID:3))) connecting to InetSocketAddress:SXXXXXXX05E/10.100.200.123:49001 with timeout:7199

 [           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) The parent thread wokeup.

 [           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)) TimeRemaining:7198; Result:SUCCESS; Max. open thread count: 2; Current open thread count:0

 [           main] c.m.s.j.i.SocketFinder                   : SocketFinder(TDSChannel (ConnectionID:3)): Closing TCP socket:Socket[addr=SXXXXXXX05E/10.100.200.123,port=49001,localport=37676]

 [           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 ClientConnectionId: 008a03a9-ba52-409e-bade-86a06e3c291b Requesting encryption level:OFF

 [           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3) read failed:Connection reset

 [           main] c.m.s.j.i.SQLServerException             : *** SQLException:ConnectionID:3 ClientConnectionId: 008a03a9-ba52-409e-bade-86a06e3c291b com.microsoft.sqlserver.jdbc.SQLServerException: Connection reset ClientCo

[           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Closing TCP socket...

 [           main] c.m.s.j.i.SQLServerConnection            : ConnectionID:3 ClientConnectionId: 008a03a9-ba52-409e-bade-86a06e3c291b Prelogin error: host SXXXXXXX05E port 49001 Error reading prelogin response: Connection

[           main] c.m.s.j.i.T.Channel                      : TDSChannel (ConnectionID:3): Closing TCP socket...

 

**All the three times Offline IP is chosen above.**

**Could you please help us based on the observations .**

**Comment by peterbae** [comment]
Hi @udaykiranraju1983, I'm going over the MSDN document [here](https://docs.microsoft.com/en-us/sql/connect/jdbc/setting-the-connection-properties?view=sql-server-ver15) and it looks like `multiSubnetFailover` should be set to by default for the driver version you're using. However, I'm a bit confused with the relationship between `multiSubnetFailover` and `transparentNetworkIPResolution` (TNIR) because with the change to `multiSubnetFailover` being set to true by default. TNIR should be disabled by default for the driver but the documentation also says it should be enabled by default.

I think the driver is currently connecting to the first IP only instead of detecting the active socket. Could you try setting `multiSubnetFailover` to false and set transparentNetworkIPResolution to true, then see if this solves your problem? I do think this might be a problem on our driver's end, but I'd like to see if this will unblock you first.

**Comment by udaykiranraju1983** [comment]
Hi @peterbae , I also looked into the documentation [here](https://docs.microsoft.com/en-us/sql/connect/jdbc/jdbc-driver-support-for-high-availability-disaster-recovery?view=sql-server-ver15). it states that transparentNetworkIPResolution is ignored if multiSubnetFailover is true. 

I have tried with transparentNetworkIPResolution set to true and removing multiSubnetFailover , still facing the same issue.

if we look at the logs closely, driver is trying to connect both IP addresses using parallel threads, but the socket connector with inactive IP address chosen then connection reset error is thrown.  this is exactly confusing me. why is the socket with inactive IP is not timed out instead of connection reset.  Ideally socket connector with active IP should be chosen and inactive IP should be timeout.

below is the pat of the logs where it is trying to connect both IPs. (10.100.200.111 - inactive IP , 10.100.200.123 - is active IP)

 main] c.m.s.j.i.SocketFinder : SocketFinder(TDSChannel (ConnectionID:3)) Total no of InetAddresses: 2. They are: SXXXXXXX05E /10.100.200.111;SXXXXXXX05E /10.100.200.123;

[ main] c.m.s.j.i.SocketFinder : SocketFinder(TDSChannel (ConnectionID:3))Using Threading with timeout:2400

[ main] c.m.s.j.i.SocketFinder : SocketFinder(TDSChannel (ConnectionID:3)) TimeRemaining:2400; Result:UNKNOWN; Max. open thread count: 2; Current open thread count:0

**[pool-6-thread-1] c.m.s.j.i.SocketConnector : SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3))) connecting to InetSocketAddress:SXXXXXXX05E /10.100.200.111:49001 with timeout:2400

[pool-6-thread-2] c.m.s.j.i.SocketConnector : SocketConnector:10(SocketFinder(TDSChannel (ConnectionID:3))) connecting to InetSocketAddress:SXXXXXXX05E /10.100.200.123:49001 with timeout:2400**

[pool-6-thread-1] c.m.s.j.i.SocketFinder : The following child thread is waiting for socketFinderLock:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-1] c.m.s.j.i.SocketFinder : The following child thread acquired socketFinderLock:SocketConnector:9(SocketFinder(TDSChannel (ConnectionID:3)))

[pool-6-thread-1] c.m.s.j.i.SocketFinder : The socket of the following thread has been chosen:**SocketConnector:9**(SocketFinder(TDSChannel (ConnectionID:3)))

**Comment by peterbae** [comment]
When you said this works from standalone java application deployed on Windows, correct? Is the windows app connecting to the same SQL Server as Linux app, or are they targeting two separate SQL Server instances?

**Comment by udaykiranraju1983** [comment]
Hi @peterbae , Yes,  standalone java application deployed on Windows. Both windows and linux apps are targeting the Single/same SQL server instance.

**Comment by peterbae** [comment]
Thanks. This is puzzling for me as well, since from the driver's perspective the underlying OS shouldn't affect the driver's behavior. Perhaps the driver thinks that both IPs are active and chooses the first one - but just to test it out, could you try having the 111 IP active and 123 IP inactive, and see if the driver still chooses the 111 IP socket?

**Comment by peterbae** [comment]
Hi @udaykiranraju1983, have you been able to try out my suggestion?

**Comment by udaykiranraju1983** [comment]
Hi @peterbae , I could not try the option suggested, that requires more effort across organization since this database is highly used.

Meanwhile, I am working with cloud/kubernetes/network support teams if they can figure out something about the issue. I will update here if there is any luck.

**Comment by peterbae** [comment]
Hi @udaykiranraju1983, were you able to find any other leads on this issue? Thanks for your help.

**Comment by peterbae** [comment]
Was there any updates on this item?

**Comment by udaykiranraju1983** [comment]
Hi @peterbae , Sorry for late reply. I was all over the place to resolve the issue. Finally able to find the fix for the issue with help of cloud support team.  

We have added below annotations to kubernates deployment file to fix the issue.

traffic.sidecar.istio.io/excludeOutboundPorts: "49001" where 49001 is the port of the Always on listener port.

We were able to isolate the problem to k8s by starting the network sniffer trace.

Observed that a SYN packet sent from the socket as below flow as I understand at high level.

Application --> istio(locatport) --> firewall --> database(49001).

A SYN packet sent to database received ACK with 49001 packet from istio itself as if it is reached 49001 
but it just reached istio and it should have sent ACK with local port, 
With this false ACK socket assumed it got connected then subsequent flow getting reset.
